7주차 과제 인가 구현하기

eastshine-high opened the pull request:
안녕하세요. 1차 커밋입니다.

구현내용
7장 Spring Security 강의 코드 적용

미구현 내용
요청 사용자에 따른 인가처리
암호화

eastshine-high committed:
7주차 과제 인가 구현하기

구현내용
7장 Spring Security 강의 코드 적용

미구현 내용
요청 사용자에 따른 인가처리
암호화

johngrib commented on app/build.gradle:
```diff
@@ -69,6 +69,9 @@ dependencies {
     runtimeOnly 'com.h2database:h2'
 
+    // Spring Security
+    implementation 'org.springframework.boot:spring-boot-starter-security'
```
spring-boot-starter-security를 소개하는 글을 작성해 보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/config/SecurityJavaConfig.java:
```diff
@@ -0,0 +1,42 @@
+
+        Filter authenticationErrorFilter = new AuthenticationErrorFilter();
+
+        http
+                .csrf().disable()
+                .addFilter(authenticationFilter)
+                .addFilterBefore(authenticationErrorFilter,
+                        JwtAuthenticationFilter.class)
+                .sessionManagement()
+                .sessionCreationPolicy(SessionCreationPolicy.STATELESS) // Rest Api를 위한 설정
+                .and()
+                .exceptionHandling()
+                .authenticationEntryPoint(
+                        new HttpStatusEntryPoint(HttpStatus.UNAUTHORIZED));
```
여기에 작성하신 설정 코드 한 줄 한 줄에 대해 설명해 보세요. 애매하게 아는 것이 있다면 글로 정리해 보시고요. 꽤 까다로운 작업이지만 좋은 공부가 될 거에요.

johngrib commented on app/src/main/java/com/codesoom/assignment/config/WebJavaConfig.java:
```diff
@@ -13,6 +13,7 @@
     @Override
     public void addInterceptors(InterceptorRegistry registry) {
-        registry.addInterceptor(authenticationInterceptor);
+        // 인터셉터를 추가해 준다. 사용 안 함. remind용.
```
이런 주석은 내용을 다듬어서 javadoc으로 재작성해보세요.

eastshine-high replied:
이렇게 추가해 보았습니다.
```
    /**
     * 인터셉터 레지스트에 권한 인증 로직을 추가한다.
     *
     * @param registry 인터셉터 레지스트
     */
```

johngrib commented on app/src/main/java/com/codesoom/assignment/filters/AuthenticationErrorFilter.java:
```diff
@@ -0,0 +1,26 @@
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+public class AuthenticationErrorFilter extends HttpFilter {
```
`HttpFilter`의 기능과 역할에 대해 설명하는 글을 작성해 보세요. 슬랙에도 공유해보시면 좋은 경험이 될 거에요.

johngrib commented on app/src/main/java/com/codesoom/assignment/interceptors/AuthenticationInterceptor.java:
```diff
@@ -61,4 +61,8 @@ private boolean doAuthentication(HttpServletRequest request,
         return true;
     }
+    // 인터셉터 결론
+    // 인터셉터를 이용하면 스프링 컨트롤러에서 처리되는 애들을 좀 모아서 깔끔하게 정리할 수 있다.
```
'처리되는 애' 가 무엇인지에 대해 좀 더 명확한 단어가 무엇일지 고민해 보세요. 짧은 주석이라도 공부할 거리를 찾아 봅시다.

johngrib reviewed (commented):
오랜만입니다. 이번주도 화이팅!

eastshine-high committed:
코드숨 7주차 강의 암호화 적용

구현 내용
코드숨 7주차 강의 암호화 적용

미구현 내용
요청 사용자에 따른 인가처리
1차 커밋 코멘트에 대한 피드백 작성

eastshine-high commented:
안녕하세요 2차 커밋입니다.

구현 내용
코드숨 7주차 강의 암호화 적용

미구현 내용
요청 사용자에 따른 인가처리

1차 커밋 코멘트에 대한 피드백도 작성하도록 하겠습니다!

johngrib commented on app/src/test/java/com/codesoom/assignment/domain/UserTest.java:
```diff
@@ -31,12 +51,11 @@ void destroy() {
 
-        assertThat(user.authenticate("test")).isTrue();
-        assertThat(user.authenticate("xxx")).isFalse();
+        assertThat(user.authenticate("test", passwordEncoder)).isTrue();
```
윗줄에 있는 `"test"`와 같은 문자열을 쓴다면 변수로 선언해서 사용해 주세요. 이런 곳에 오타가 한번 들어가면 테스트가 깨지겠죠. 지금은 테스트 코드가 짧아서 둘 사이의 거리도 가깝지만, 테스트 코드가 꽤 복잡해져서 위의 "test"와 아래의 "test" 사이에 15줄 정도 들어가면 문제가 생겼을 때 원인을 힘겹게 찾게 될 수도 있습니다.

eastshine-high committed:
3차 커밋

구현 내용
상품 정보 생성 시 작성자 정보 추가
상품 정보 수정 시 수정 권한이 있는 지를  확인

미구현 내용
구현한 로직의 테스트 코드 작성(TDD를 적용하지 않았습니다....)

eastshine-high commented:
3차 커밋입니다.

구현 내용
* 상품 정보 생성 시 작성자 정보 추가 (JPA 연관관계 매핑을 공부하여 적용해 보았습니다)
* 상품 정보 수정 시 수정 권한이 있는 지를  확인

미구현 내용
* 구현한 로직의 테스트 코드 작성(TDD를 적용하지 않았습니다....)

부족함을 많이 느끼네요.

johngrib commented on app/src/main/java/com/codesoom/assignment/application/ProductService.java:
```diff
@@ -32,14 +36,29 @@ public Product getProduct(Long id) {
+     * @param productData 수정할 상품의 목표 정보
+     * @param authentication 권한 정보
+     * @return 수정된 상품 정보
+     * @Throw 권한이 없을 경우
```
어떤 예외 타입을 던지는지 적어주도록 합시다.
```suggestion
     * @Throw InvalidRoleException 권한이 없을 경우
```

johngrib commented on app/src/main/java/com/codesoom/assignment/config/SecurityJavaConfig.java:
```diff
@@ -0,0 +1,51 @@
+
+    @Override
+    public void configure(WebSecurity web) throws Exception {
+        web.ignoring().antMatchers("/h2-console/**");
```
`ignoring`이 어떤 용도로 사용되는지 자세히 설명해 보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/security/UserAuthentication.java:
```diff
@@ -0,0 +1,50 @@
+    private static List<GrantedAuthority> authorities(Long userId) {
+        List<GrantedAuthority> authorities = new ArrayList<>();
+
+        // Todo 도메인 설계 필요
+        if(userId == 1L){
+            authorities.add(new SimpleGrantedAuthority("ADMIN"));
+        }else{
+            authorities.add(new SimpleGrantedAuthority("USER"));
+        }
+        return authorities;
```
이 `if`문은 조금 더 보기 좋게 개선할 수 있을 것 같아요.

eastshine-high replied:
다음과 같이 하면 괜찮을까요?
```
if(userId == 1L) authorities.add(new SimpleGrantedAuthority("ADMIN"));
else authorities.add(new SimpleGrantedAuthority("USER"));
```

johngrib replied:
`String`에 대해서만 `if`를 써도 괜찮겠죠.

```
authorities.add(userId == 1L ? "ADMIN" : "USER");
```

다만 `1L`이 있기 때문에 더 나은 방법을 열심히 고민해봐야 할 거에요.

johngrib commented:
> 부족함을 많이 느끼네요.

멘탈이 힘드신가 보군요. 과제 달성에 너무 힘을 쏟고 있어서 피로를 느끼신다면, 잠시 과제를 멈추고 지난주나 지지난주 과제를 다시 한번 해보시는 것도 권해드려요. 2주차 것도 괜찮습니다. 쉬운 것을 해보면서 자신감을 다시 찾은 다음 과제에 재도전해보는 것도 나쁘지 않아요.

eastshine-high commented:
> > 부족함을 많이 느끼네요.
> 
> 멘탈이 힘드신가 보군요. 과제 달성에 너무 힘을 쏟고 있어서 피로를 느끼신다면, 잠시 과제를 멈추고 지난주나 지지난주 과제를 다시 한번 해보시는 것도 권해드려요. 2주차 것도 괜찮습니다. 쉬운 것을 해보면서 자신감을 다시 찾은 다음 과제에 재도전해보는 것도 나쁘지 않아요.

조언 감사합니다.
오늘은 커밋을 하지 않고 스스로가 부족하다고 생각되는 부분들을 학습하였습니다.
이번 한 주도 고생하셧습니다!

johngrib commented:
어느새 일요일이 되었군요. 고생 많으셨습니다. 주간 회고를 작성해 공유해 주세요.
